### ISSUE TITLE
Freya.TodoBackend no longer passes todobackend.com tests

### ISSUE THREAD

**Issue by panesofglass** [freya-fs/freya.examples#4]

**Comment by panesofglass** [comment]
Chrome log shows:

```
XMLHttpRequest cannot load http://localhost:7000/.... No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:8083' is therefore not allowed access.
```

**Comment by kolektiv** [comment]
Right, so... Discovered a threading/race error of some kind with async streams which caused some problems initially. In the prelude for TodoBackend, change the body helper section to the following:

```fsharp
(* Request Body Helper

   Freya doesn't provide built-in ways of extracting data from the body of
   a request, as it's usually very specific to an application, and the Freya
   way is to let the developer choose the most suitable approach.

   We've used Chiron in this example, so we can use that to define the body
   function below, which (following from Chiron) uses static inference to
   determine the type of return value needed. *)

let readBody =
    freya {
        let! stream = Freya.Optic.get Request.body_

        use reader = new StreamReader (stream)
        return reader.ReadToEnd () }

let inline body () =
    freya {
        let! body = readBody
        return (Json.tryParse body |> Either.bind Json.tryDeserialize |> Either.toOption) }
```

With that done, the TodoBackend works perfectly on Katana. However, even on Windows, it doesn't on Suave (same code, just different server). So that does make me think that Suave just isn't doing anything with those headers (CORS). I'm not sure how the internal bridging between OWIN and Suave works, but it's not a direct mapping I know, so perhaps Suave isn't set up for those headers in some way? cc. @haf for an opinion...?

**Comment by kolektiv** [comment]
Saying that, writing that header manually seems more successful, so it could be an initial parsing phase issue...

**Comment by kolektiv** [comment]
It does seem to be down to the implementation of TryGetValue on the DeltaDictionary in Suave. The value returned from the DeltaDictionary is null when you use the pattern match syntax on it:

```fsharp
match dict.TryGetValue key with
| true, value -> Some value
| _,_ -> None
```

will result in the value being null even when it isn't null in the dictionary. This means that when using a delta dictionary for the request headers, Freya can't read request headers in Suave.

**Comment by kolektiv** [comment]
cc. @haf on this one...

**Comment by kolektiv** [comment]
Replacing the TryGetValue on the DeltaDictionary in Suave with the following:

```fsharp
member x.TryGetValue (key, [<Out>] res : byref<string[]>) =
  match Map.tryFind (OwinKey key) x.Delta with
  | Some v -> res <- v; true
  | _ -> false
```

...fixes the problem, seemingly. This might actually be a bug in the F# map when cast as an IDictionary<_,_>...

**Comment by catlion** [comment]
I believe I've seen this error occasionally in NancyFx on Windows too. But I've already migrated to Freya, so can't reproduce.

Also, Isn't `Interlocked.Exchange(&res, v)` better?

**Comment by haf** [comment]
So I think I've released a fix for this now, v1.1.1 of Suave.

**Comment by kolektiv** [comment]
The Suave release fixed the problem with the header dictionary!

But... I'm now running up against this: https://github.com/SuaveIO/suave/issues/387
